### PR TITLE
Fix #331 "We only get `Const::Num`, never `Const::Int`"

### LINKED ISSUE #331: We only get `Const::Num`, never `Const::Int`
This is a bug in the lexer.

Currently we only have one numerical token type `NumericalLiteral` which is of type `f64`, hence all numerical values are converted to `f64` and we only get `Cont::Num`.

This bug blocks #328

### PR DESCRIPTION
**Pull request by HalidOdat**
This will close #331 and unblock #328

### CODE REVIEW

**Comment by jasonwilliams** [comment]
Thanks, will try and get round to looking at this soon

File: boa/src/syntax/lexer/mod.rs [hunk 1]
```diff
@@ -399,16 +399,18 @@ impl<'a> Lexer<'a> {
                                     }
                                 }
                             }
+                            dbg!(&gone_decimal);
```
Comment (Razican) [inline boa/src/syntax/lexer/mod.rs:402]: This should be removed before merging.

File: boa/src/syntax/lexer/mod.rs [hunk 2]
```diff
@@ -399,16 +399,18 @@ impl<'a> Lexer<'a> {
...
                             if gone_decimal {
-                                f64::from_str(&buf).map_err(|_e| LexerError::new("Could not convert value to f64"))?
+                                NumericLiteral::Number(f64::from_str(&buf).map_err(|_e| LexerError::new("Could not convert value to f64"))?)
                             } else if buf.is_empty() {
-                                0.0
+                                NumericLiteral::Integer(0)
                             } else {
-                                (u64::from_str_radix(&buf, 8).map_err(|_e| LexerError::new("Could not convert value to u64"))?) as f64
+                                NumericLiteral::Integer(
+                                u64::from_str_radix(&buf, 8).map_err(|_e| {dbg!(&_e); LexerError::new("Could not convert value to u64")})?)
```
Comment (Razican) [inline boa/src/syntax/lexer/mod.rs:409]: The `dbg!()` call should be removed before merging.
```suggestion
                                u64::from_str_radix(&buf, 8).map_err(|_e| LexerError::new("Could not convert value to u64"))?)
```

File: boa/src/syntax/lexer/mod.rs [hunk 3]
```diff
@@ -469,9 +474,19 @@ impl<'a> Lexer<'a> {
...
                     self.push_token(TokenKind::NumericLiteral(
-                        f64::from_str(&buf).map_err(|_| LexerError::new("Could not convert value to f64"))?,
+                        if is_floating_point {
+                            // TODO make this a bit more safe -------------------------------VVVV
+                            NumericLiteral::Number(
+                                f64::from_str(&buf).map_err(|_| LexerError::new("Could not convert value to f64"))?
+                            )
+                        } else {
+                            NumericLiteral::Integer(
+                                u64::from_str(&buf).map_err(|_e| {dbg!(&_e); dbg!(&is_floating_point); LexerError::new("Could not convert value to u64")})?
```
Comment (Razican) [inline boa/src/syntax/lexer/mod.rs:486]: The `dbg!()` macro should be removed before merging.
```suggestion
                                u64::from_str(&buf).map_err(|_e| LexerError::new("Could not convert value to u64"))?
```

File: boa/src/syntax/ast/token.rs [hunk 4]
```diff
@@ -42,6 +42,17 @@ impl Debug for VecToken {
     }
 }
 
+/// Represents the type differenct types of numeric literals.
+#[cfg_attr(feature = "serde-ast", derive(Serialize, Deserialize))]
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum NumericLiteral {
+    /// A floating point number
+    Number(f64),
```
Comment (Razican) [inline boa/src/syntax/ast/token.rs:50]: I would probably change this `Number` for `Real` or something like that, to differentiate from an `Integer`.

File: boa/src/syntax/ast/token.rs [hunk 5]
```diff
@@ -42,6 +42,17 @@ impl Debug for VecToken {
...
+    /// A integer
+    Integer(u64),
+    // TODO: Add BigInt
```
Comment (Razican) [inline boa/src/syntax/ast/token.rs:53]: How big is a `BigInt`?

Comment (HalidOdat) [reply boa/src/syntax/ast/token.rs:53]: BigInt can hold any arbitrary large integer. All the numerical are implemented in the software level.

[num-bigint](https://crates.io/crates/num-bigint) crate

File: boa/src/syntax/ast/token.rs (hunk 4 above)
Comment (HalidOdat) [reply boa/src/syntax/ast/token.rs:54]: In V8 they have `Number` and `Smi`, so should I do it like V8 or `Real` and `Integer`?

Comment (Razican) [reply boa/src/syntax/ast/token.rs:54]: > In V8 they have `Number` and `Smi`, so should I do it like V8 or `Real` and `Integer`?

I think it's more accurate if we call `Real` or even `Rational` to non-integers.

**Comment by HalidOdat** [comment]
I reworked the number lexing and fixed some things:
 - Added some documentation to `lexer` and `token`
 - Fixed a bug were it did not accept `09.5e2`.
 - Made it easy to add `strict mode` for number literal (in strict mode `034` and `098` is an error).
 - Made it easy to add `BigInt` support (including other base BigInts).
 - Added `Value::is_integer` to check if a number is an integer.
 - Renamed `ValueData::Number(f64)` to `ValueData::Racional(f64)`
 - Fixed position bug related to numeric tokens.
 - Added some tests for edge case with implicit octal and hexadecimals followed by a dot
 - Split numerical lexing to a different function.
 - Removed `read_integer_in_base`.
 
This **unblocks** #328 

I think now it's ready to review. :)

File: boa/src/builtins/value/mod.rs [hunk 6]
```diff
@@ -43,7 +43,7 @@ pub enum ValueData {
     /// `String` - A UTF-8 string, such as `"Hello, world"`
     String(String),
     /// `Number` - A 64-bit floating point number, such as `3.1415`
-    Number(f64),
+    Racional(f64),
```
Comment (Razican) [inline boa/src/builtins/value/mod.rs:46]: Why is this `Racional` instead of `Rational`?

File: boa/src/syntax/lexer/tests.rs [hunk 7]
```diff
@@ -336,7 +336,6 @@ fn test_two_divisions_in_expression() {
     let s = "    return a !== 0 || 1 / a === 1 / b;";
     let mut lexer = Lexer::new(s);
     lexer.lex().expect("failed to lex");
-    dbg!(&lexer.tokens);
```
Comment (Razican) [inline boa/src/syntax/lexer/tests.rs:339]: This `dbg!` here was to be able to debug the token generation from the lexer when we activate this test (when we solve the goal-oriented lexer). We can remove it, but not sure if there is a need.

File: boa/src/syntax/ast/token.rs [hunk 8]
```diff
@@ -42,28 +46,53 @@ impl Debug for VecToken {
     }
 }
 
+/// Represents the type differenct types of numeric literals.
+#[cfg_attr(feature = "serde-ast", derive(Serialize, Deserialize))]
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum NumericLiteral {
+    /// A floating point number
+    Racional(f64),
```
Comment (Razican) [inline boa/src/syntax/ast/token.rs:54]: Again, isn't the correct word `Rational`?

File: boa/src/syntax/lexer/tests.rs [hunk 9]
```diff
@@ -443,55 +541,88 @@ fn test_regex_literal_flags() {
...
-    assert_eq!(lexer.tokens[0].kind, TokenKind::NumericLiteral(1000.0));
+    assert_eq!(
+        lexer.tokens[0].kind,
+        TokenKind::NumericLiteral(NumericLiteral::Racional(1000.0))
+    );
     assert_eq!(lexer.tokens[1].kind, TokenKind::Punctuator(Punctuator::Add));
     assert_eq!(
         lexer.tokens[2].kind,
-        TokenKind::NumericLiteral(100_000_000_000.0)
+        TokenKind::NumericLiteral(NumericLiteral::Racional(100_000_000_000.0))
```
Comment (Razican) [inline boa/src/syntax/lexer/tests.rs:626]: If we start seeing this a bit too wordy, we could implement `From<f64>` and `From<i32>` for `NumericLiteral`, and then add a `fn numeric_literal<L>(lit: L) -> Self where L: Into<NumericLiteral> {}` to `TokenKind`, which would directly allow us to pass a `f64` or a `i32`. I think I already did something similar in the modular parser PR.

File: boa/src/builtins/value/mod.rs (hunk 6 above)
Comment (HalidOdat) [reply boa/src/builtins/value/mod.rs:46]: > Why is this `Racional` instead of `Rational`?

Good catch! I wouldn't have noticed if I wasn't concentrated on it. It's because in my native language that's
how you spell `Rational` in my mind it didn't seem wrong at the time.

You're right it should be `Rational`. :)

File: boa/src/syntax/lexer/mod.rs [hunk 10]
```diff
@@ -229,6 +223,225 @@ impl<'a> Lexer<'a> {
...
+                *self = match *self {
+                    Self::Rational => unreachable!("can not convert rational number to BigInt"),
+                    Self::Integer(base) => Self::BigInt(base),
+                    Self::BigInt(base) => Self::BigInt(base),
+                };
+            }
+        }
+
+        // TODO: Setup strict mode.
+        let strict_mode = false;
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/mod.rs:263]: Currently I have `strict_mode` set to always be `false`. In strict mode decimals with leading zeros and implicit octal are not allowed.

I seems like the parser should take in a lexer, because `'use strict';` it only applies to function or program scope and the lexer does not know what a scope is.

File: boa/src/builtins/value/mod.rs (hunk 6 above)
Comment (Razican) [reply boa/src/builtins/value/mod.rs]: It happens to me too. In Spanish it's spelled "racional" too haha

File: boa/src/syntax/lexer/mod.rs (hunk 10 above)
Comment (Razican) [reply boa/src/syntax/lexer/mod.rs:263]: Yep, this is going to be interesting to see once we have the lexer as an iterator tied to the cursor.

**Comment by Razican** [comment]
I would wait for #304 to land before this.

**Comment by HalidOdat** [comment]
I will rebase today!
